File: 109-Convert SortedListtoBinarySearchTree-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    static ListNode currentHead=null;
    public TreeNode toBST(int start, int end) {
        if (start>end) return null;
        int mid=start+(end-start)/2;
        TreeNode left=toBST(start,mid-1);
        TreeNode head=new TreeNode(currentHead.val);
        currentHead=currentHead.next;
        head.left=left;
        head.right=toBST(mid+1,end);
        return head;
    }
    public TreeNode sortedListToBST(ListNode head) {
        int n=0;
        ListNode node=head;
        while (node!=null) {
            n++;
            node=node.next;
        }
        currentHead=head;
        return toBST(0,n-1);
    }
}

File: 109-Convert SortedListtoBinarySearchTree-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* toBST(ListNode* &head, int start, int end) {
        if (start>end) return NULL;
        int mid=start+(end-start)/2;
        TreeNode* left=toBST(head,start,mid-1);
        TreeNode* treeHead=new TreeNode(head->val);
        head=head->next;
        TreeNode* right=toBST(head,mid+1,end);
        treeHead->left=left;
        treeHead->right=right;
        return treeHead;
    }
    TreeNode* sortedListToBST(ListNode* head) {
        int n=0;
        ListNode* node=head;
        while (node) {
            n++;
            node=node->next;
        }
        return toBST(head,0,n-1);
    }
};
